app/components/Common/PremiumPaywall.tsx

```diff
@@ -0,0 +1,142 @@
+import React, { useState, useEffect } from 'react'
+import { useNavigate } from '@remix-run/react'
+import DiscordIcon from '~/icons/DiscordIcon'
+import { refreshDiscordRoles, DISCORD_SERVER_URL } from '~/utils/premium'
+
+interface PremiumPaywallProps {
+  loaderData: {
+    isLoggedIn: boolean
+    hasPremium: boolean
+    needsRefresh?: boolean
+  }
+  children: React.ReactNode
+}
+
+const PremiumPaywall: React.FC<PremiumPaywallProps> = ({
+  loaderData,
+  children
+}) => {
+  const [isRefreshing, setIsRefreshing] = useState(false)
+  const navigate = useNavigate()
+
+  const { isLoggedIn, hasPremium, needsRefresh = false } = loaderData
+
+  // Determine if paywall should be shown
+  const showPaywall = !isLoggedIn || !hasPremium || needsRefresh
+
+  // Automatically refresh roles when needsRefresh is true and user is logged in
+  useEffect(() => {
+    if (isLoggedIn && needsRefresh && !isRefreshing) {
+      const refreshRoles = async () => {
+        await handleRefresh()
+      }
+      refreshRoles()
+    }
+  }, [needsRefresh, isLoggedIn])
+
+  const handleRefresh = async () => {
+    if (isRefreshing) return
+
+    setIsRefreshing(true)
+    try {
+      await refreshDiscordRoles()
+    } catch (error) {
+      console.error('Failed to refresh roles:', error)
+      // Fallback to page reload
+      window.location.reload()
+    } finally {
+      setIsRefreshing(false)
+    }
+  }
+
+  const handleLogin = () => {
+    navigate('/discord-login')
+  }
+
+  const handleSubscribe = () => {
+    window.open(DISCORD_SERVER_URL, '_blank')
+  }
+
+  if (!showPaywall) return <>{children}</>
+
+  return (
+    <div className="relative">
+      <div className="pointer-events-none filter blur-sm select-none opacity-60">
+        {children}
```

Comment:
This will expose the premium content to anyone capable of using the inspector tool to remove the absolutely positioned element.

Reply:
https://github.com/ff14-advanced-market-search/saddlebag-with-pockets/pull/649/commits/914007dc1bdc8bbb74ce2824c18e571139268d5f